File: haskell/p-27.hs
import qualified Data.List as D

combinations :: (Eq a, Num a) => a -> [t] -> [[t]]
combinations k [] = []
combinations 1 (x:xs) = [[x]] ++ (combinations 1 xs)
combinations k (x:xs) = (map (\ l -> x:l) (combinations (k - 1) xs)) ++
                        (combinations k xs)

myGroup [] [] = [[]]
myGroup (n:ns) xs =
    [x : y | x <- combinations n xs, y <- myGroup ns (xs D.\\ x)]

-- *Main> length $ myGroup [2,3,4] ["aldo","beat","carla","david","evi","flip","gary","hugo","ida"]
-- 1260
-- *Main> length $ myGroup [2,2,5] ["aldo","beat","carla","david","evi","flip","gary","hugo","ida"]
-- 756

File: erlang/p_27.erl
-module(p_27).
-export([my_group/2]).
-import(p_26, [combinations/2]).

my_group([], []) ->
    [[]];
my_group([N|Ns], Xs) ->
    [Y ++ Ys || Y <- combinations(N, Xs),
		Ys <- my_group(Ns, Xs -- Y)].

%% 121> length(p_27:my_group([2,3,4], [a,b,c,d,e,f,g,h,i])).
%% 1260
%% 122> length(p_27:my_group([2,2,5], [a,b,c,d,e,f,g,h,i])).
%% 756


